src/librustc_driver/lib.rs

```diff
@@ -1419,16 +1419,49 @@ pub fn in_rustc_thread<F, R>(f: F) -> Result<R, Box<Any + Send>>
     // Temporarily have stack size set to 16MB to deal with nom-using crates failing
     const STACK_SIZE: usize = 16 * 1024 * 1024; // 16MB
 
-    let mut cfg = thread::Builder::new().name("rustc".to_string());
+    #[cfg(unix)]
+    let spawn_thread = unsafe {
+        // Fetch the current resource limits
+        let mut rlim = libc::rlimit {
+            rlim_cur: 0,
+            rlim_max: 0,
+        };
+        if libc::getrlimit(libc::RLIMIT_STACK, &mut rlim) != 0 {
+            let err = io::Error::last_os_error();
+            panic!("in_rustc_thread: error calling getrlimit: {}", err);
+        }
 
-    // FIXME: Hacks on hacks. If the env is trying to override the stack size
-    // then *don't* set it explicitly.
-    if env::var_os("RUST_MIN_STACK").is_none() {
-        cfg = cfg.stack_size(STACK_SIZE);
-    }
+        if STACK_SIZE as libc::rlim_t > rlim.rlim_max {
+            true
+        } else {
+            rlim.rlim_cur = STACK_SIZE as libc::rlim_t;
+            // Set our newly-increased resource limit
+            if libc::setrlimit(libc::RLIMIT_STACK, &rlim) != 0 {
+                let err = io::Error::last_os_error();
+                panic!("in_rustc_thread: error calling setrlimit: {}", err);
+            }
 
-    let thread = cfg.spawn(f);
-    thread.unwrap().join()
+            false
+        }
+    };
+
+    #[cfg(not(unix))]
+    let spawn_thread = true;
+
+    if spawn_thread {
+        let mut cfg = thread::Builder::new().name("rustc".to_string());
+
+        // FIXME: Hacks on hacks. If the env is trying to override the stack size
+        // then *don't* set it explicitly.
+        if env::var_os("RUST_MIN_STACK").is_none() {
```

Comment:
Could this be checked earlier? If this env var is set I think we'll want to force spawning a thread.

Reply:
I disagree: that environment variable only affects stack size of non-main thread. If we can extend the main thread stack, that's all what we need.

Reply:
Without moving this check earlier it's breaking existing behavior. Today the stack size of the rustc thread can be controlled by configuring `RUST_MIN_STACK` (can be useful for debugging). This commit breaks that ability because setrlimit will probably succeed on Linux.